lib/resty/rediscluster.lua

```diff
@@ -726,64 +1013,88 @@ function _M.commit_pipeline(self)
     -- coroutine swich happens(eg. ngx.sleep, cosocket), very important!
     slots = nil -- luacheck: ignore
 
+    local magic_random_seed = generate_magic_seed(self)
+
     for k, v in pairs(node_req_map) do
-        local ip = v.ip
-        local port = v.port
         local reqs = v.reqs
-        local slave = v.slave
-        local redis_client = redis:new()
-        redis_client:set_timeouts(config.connect_timeout or DEFAULT_CONNECTION_TIMEOUT,
-                                  config.send_timeout or DEFAULT_SEND_TIMEOUT,
-                                  config.read_timeout or DEFAULT_READ_TIMEOUT)
-        local ok, err = redis_client:connect(ip, port, self.config.connect_opts)
+
+        local ip, port
+        local redis_client
+        local ok, err_t = nil, {}
+        -- Redis 7: try IP first and then hostname as IP does not suffer from stale DNS cache
+        local tries = redis7_with_host and 2 or 1
+        for i = 1, tries do
+            local slave, err
+            ip, port, slave, err = pick_node(self, v.serv_list, nil, magic_random_seed, i)
+            if err then
+                table_insert(err_t, err)
+                self:refresh_slots()
+                return nil, table_concat(err_t, ",")
+            end
+
+            redis_client = redis:new()
+            redis_client:set_timeouts(config.connect_timeout or DEFAULT_CONNECTION_TIMEOUT,
+                                      config.send_timeout or DEFAULT_SEND_TIMEOUT,
+                                      config.read_timeout or DEFAULT_READ_TIMEOUT)
+            ok, err = redis_client:connect(ip, port, self.config.connect_opts)
+            if ok then
+                v.slave = slave
+                break
+            end
+            table_insert(err_t, ip .. ":" .. port .. " " .. err)
+        end
         if not ok then
-            return nil, "failed to connect, err: " .. err
+            self:refresh_slots()
```

Comment:
why after connect failed need to `refresh_slots`? `refresh_slots` also need to connect redis?

Reply:
1. `refresh_slots` will try a full list of `ip:port` (including cached and configed).
2. Usually, within k8s environment, the cluster is recovered very soon. So, the configured hostname can help recover.

Reply:
`self:refresh_slots()` -> `self:fetch_slots()` -> `try_hosts_slots` -> `redis_client:connect` 

since connect failed, `self:refresh_slots()` will eventually execute `redis_client:connect`, is this a failure loop?

Reply:
This function will loop through the servers list, to recover. For each server, a new `redis_client` is created.

```
local _, errors = try_hosts_slots(self, serv_list_combined)
```